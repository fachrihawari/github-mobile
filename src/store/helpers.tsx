import { Action as ReduxAction } from "redux";

export interface Action<T extends string, P> extends ReduxAction<T> {
  payload: P;
}

export function createAction<T extends string>(type: T): Action<T, void>;
export function createAction<T extends string, P>(
  type: T,
  payload: P
): Action<T, P>;
export function createAction<T extends string, P>(type: T, payload?: P) {
  return typeof payload === "undefined" ? { type } : { type, payload };
}
