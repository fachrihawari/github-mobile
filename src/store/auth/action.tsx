import { FETCH_USER_REQUEST, LOGOUT } from "./constant";
import { createAction } from "../helpers";

export const fetchUserRequest = (username: string, password: string, OTP: string='') =>
  createAction(FETCH_USER_REQUEST, { username, password, OTP });

export const logout = () => createAction(LOGOUT);
