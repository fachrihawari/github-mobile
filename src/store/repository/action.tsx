import { FETCH_COMMIT_REQUEST } from "./constant";
import { createAction } from "../helpers";

export const fetchCommitRequest = (
  repository: string,
  page?: number,
  perPage?: number
) => createAction(FETCH_COMMIT_REQUEST, { repository, page, perPage });
