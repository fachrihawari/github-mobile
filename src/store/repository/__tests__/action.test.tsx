import { parseError } from '../../helpers';
import * as action from '../action';
import * as constant from '../constant';

describe('repository actions', () => {
  it('should create an action to fetch commit', () => {
    const repository = 'facebook/react-ative'
    const fetchUserRequestAction = action.fetchCommitRequest(repository)
    const expectedAction = {
      type: constant.FETCH_COMMIT_REQUEST,
      payload: {
        repository,
        page: undefined,
        perPage: undefined
      }
    }
    expect(fetchUserRequestAction).toStrictEqual(expectedAction)
    expect(fetchUserRequestAction).toMatchSnapshot()
  })

  it('should create an action to fetch commit to another page', () => {
    const repository = 'facebook/react-ative'
    
    let fetchUserRequestAction = action.fetchCommitRequest(repository, 1, 10)
    const expectedAction = {
      type: constant.FETCH_COMMIT_REQUEST,
      payload: {
        repository,
        page: 1,
        perPage: 10
      }
    }
    expect(fetchUserRequestAction).toStrictEqual(expectedAction)
    expect(fetchUserRequestAction).toMatchSnapshot()
    
    fetchUserRequestAction = action.fetchCommitRequest(repository, 2, 10)
    expectedAction.payload.page = 2
    expect(fetchUserRequestAction).toStrictEqual(expectedAction)
    expect(fetchUserRequestAction).toMatchSnapshot()
  })

  it('should create an action to set loader when fetching user', () => {
    const expectedAction = {
      type: constant.FETCH_COMMIT_BEGIN
    }
    expect(action.fetchCommitBegin()).toStrictEqual(expectedAction)
    expect(action.fetchCommitBegin()).toMatchSnapshot()
  })

  it('should create an action to set commits when success fetching commits', () => {
    const expectedAction = {
      type: constant.FETCH_COMMIT_SUCCESS,
      payload: {
        commits: [],
        currentPage: 1
      }
    }
    expect(action.fetchCommitSuccess([], 1)).toStrictEqual(expectedAction)
    expect(action.fetchCommitSuccess([], 1)).toMatchSnapshot()
  })

  it('should create an action to set error message when failure fetching commits', () => {
    const error = {
      response: {
        data: {
          message: 'Not Found'
        }
      }
    }
    const expectedAction = {
      type: constant.FETCH_COMMIT_FAILURE,
      payload: {
        error: parseError(error)
      }
    }
    expect(action.fetchCommitFailure(parseError(error))).toStrictEqual(expectedAction)
    expect(action.fetchCommitFailure(parseError(error))).toMatchSnapshot()
  })

  it('should create an action to set error message when failure fetching commits because network problem', () => {
    const error = 'Lost connection'
    const expectedAction = {
      type: constant.FETCH_COMMIT_FAILURE,
      payload: {
        error: parseError(error)
      }
    }
    expect(action.fetchCommitFailure(parseError(error))).toStrictEqual(expectedAction)
    expect(action.fetchCommitFailure(parseError(error))).toMatchSnapshot()
  })
})