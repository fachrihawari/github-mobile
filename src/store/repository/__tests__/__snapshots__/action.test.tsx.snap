// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`repository actions should create an action to fetch commit 1`] = `
Object {
  "payload": Object {
    "page": undefined,
    "perPage": undefined,
    "repository": "facebook/react-ative",
  },
  "type": "repository/FETCH_COMMIT_REQUEST",
}
`;

exports[`repository actions should create an action to fetch commit to another page 1`] = `
Object {
  "payload": Object {
    "page": 1,
    "perPage": 10,
    "repository": "facebook/react-ative",
  },
  "type": "repository/FETCH_COMMIT_REQUEST",
}
`;

exports[`repository actions should create an action to fetch commit to another page 2`] = `
Object {
  "payload": Object {
    "page": 2,
    "perPage": 10,
    "repository": "facebook/react-ative",
  },
  "type": "repository/FETCH_COMMIT_REQUEST",
}
`;

exports[`repository actions should create an action to set commits when success fetching commits 1`] = `
Object {
  "payload": Object {
    "commits": Array [],
    "currentPage": 1,
  },
  "type": "repository/FETCH_COMMIT_SUCCESS",
}
`;

exports[`repository actions should create an action to set error message when failure fetching commits 1`] = `
Object {
  "payload": Object {
    "error": "Not Found",
  },
  "type": "repository/FETCH_COMMIT_FAILURE",
}
`;

exports[`repository actions should create an action to set error message when failure fetching commits because network problem 1`] = `
Object {
  "payload": Object {
    "error": "You're offline!",
  },
  "type": "repository/FETCH_COMMIT_FAILURE",
}
`;

exports[`repository actions should create an action to set loader when fetching user 1`] = `
Object {
  "type": "repository/FETCH_COMMIT_BEGIN",
}
`;
